### ISSUE TITLE
1.1GA/nightly: NON_NEGATIVE_DERIVATIVE() not returning negative derivatives at all (rather than returning the absolute value)

### ISSUE THREAD

**Issue by daviesalex** [influxdata/influxdb#7654]
This may be a misunderstanding on my part, but I expected NON_NEGATIVE_DERIVATIVE()  to be basically the absolute value of DERIVATIVE(), i.e. if the derivative is negative return a positive integer. What actually appears to be happening is that negative derivatives are silently not returned.

Here is a test case using the API:

```
import requests, urllib, pprint, dateutil.parser, datetime

def get_timestamp(dt):
    dt = dateutil.parser.parse(dt)
    utc_naive  = dt.replace(tzinfo=None) - dt.utcoffset()
    timestamp = (utc_naive - datetime.datetime(1970, 1, 1)).total_seconds()
    return timestamp

query = """ SELECT NumRecv FROM metricname WHERE "a"='b' AND time > now() - 2m;  """
r = requests.get("http://host:8086/query?%s", urllib.urlencode({ "db": "tg_udp", "q": query}))

last_dt = 0 
for i in  r.json()["results"][0]["series"][0]["values"]:
    dt=i[0]
    delta = get_timestamp(dt) - last_dt
    if delta < 1: continue
    if delta >= 12 and last_dt != 0:
        print "** More than 12 second gap (%s) **" % (delta)
    last_dt = get_timestamp(dt)
    print "%s %s" % (dateutil.parser.parse(dt), i[1])
    
print "TEST 2" 

query = """ SELECT NON_NEGATIVE_DERIVATIVE(NumRecv, 10s) FROM metricname WHERE "a" = 'b' AND time > now() - 2m"""
r = requests.get("http://host:8086/query?%s", urllib.urlencode({ "db": "tg_udp", "q": query}))

last_dt = 0 
for dt,value in  r.json()["results"][0]["series"][0]["values"]:
    delta = get_timestamp(dt) - last_dt
    if delta >= 12 and last_dt != 0:
        print "** More than 12 second gap (%s) **" % (delta)
    last_dt = get_timestamp(dt)
    print "%s %s" % (dateutil.parser.parse(dt), value)

print "TEST 3" 
    
query = """ SELECT DERIVATIVE(NumRecv, 10s) FROM metricname WHERE "a"='b' AND time > now() - 2m"""
r = requests.get("http://host:8086/query?%s", urllib.urlencode({ "db": "tg_udp", "q": query}))

last_dt = 0 
for dt,value in  r.json()["results"][0]["series"][0]["values"]:
    delta = get_timestamp(dt) - last_dt
    if delta >= 12 and last_dt != 0:
        print "** More than 12 second gap (%s) **" % (delta)
    last_dt = get_timestamp(dt)
    print "%s %s" % (dateutil.parser.parse(dt), value)
```

Results:

```
2016-11-24 20:05:00+00:00 1764176
2016-11-24 20:05:10+00:00 432386
2016-11-24 20:05:20+00:00 3182623
2016-11-24 20:05:30+00:00 2062197
2016-11-24 20:05:40+00:00 432542
2016-11-24 20:05:50+00:00 113689
2016-11-24 20:06:00+00:00 65630
2016-11-24 20:06:10+00:00 113694
2016-11-24 20:06:20+00:00 47470
2016-11-24 20:06:30+00:00 1112988
2016-11-24 20:06:40+00:00 432400
2016-11-24 20:06:50+00:00 827283
TEST 2
2016-11-24 20:05:20+00:00 2750237.0
** More than 12 second gap (50.0) **
2016-11-24 20:06:10+00:00 48064
** More than 12 second gap (20.0) **
2016-11-24 20:06:30+00:00 1065518.0
** More than 12 second gap (20.0) **
2016-11-24 20:06:50+00:00 394883
TEST 3
2016-11-24 20:05:10+00:00 -1331790.0
2016-11-24 20:05:20+00:00 2750237.0
2016-11-24 20:05:30+00:00 -1120426.0
2016-11-24 20:05:40+00:00 -1629655.0
2016-11-24 20:05:50+00:00 -318853
2016-11-24 20:06:00+00:00 -48059
2016-11-24 20:06:10+00:00 48064
2016-11-24 20:06:20+00:00 -66224
2016-11-24 20:06:30+00:00 1065518.0
2016-11-24 20:06:40+00:00 -680588
2016-11-24 20:06:50+00:00 394883
```

It seems that the only positive values (2750237.0, 48064, 1065518.0, 394883) are returned. I expected it to return all the values, just with the negative integers converted to positive (1331790.0, 2750237.0, 1120426.0, 1629655.0 etc.)

I can reproduce this inside the shell (the timestamps suck, but the number of values shows the problem):

```
> SELECT NumRecv FROM metricname WHERE "a" = 'b' AND time > now() - 2m;
name: metricname
time                    NumRecv
----                    -------
1480018080000000000     1764455
1480018080000000000     432570
1480018080000000000     34716
1480018080000000000     827323
1480018080000000000     18996
1480018080000000000     65646
1480018080000000000     113724
1480018080000000000     432428
1480018080000000000     1113032
1480018080000000000     1009673
1480018080000000000     46671
1480018090000000000     47498
1480018090000000000     340713
1480018090000000000     2063426
1480018090000000000     34719
1480018090000000000     432572
1480018090000000000     136743
1480018090000000000     18998
1480018090000000000     3184893
1480018090000000000     46671
1480018100000000000     432431
1480018100000000000     1764479
1480018100000000000     47500
1480018100000000000     34721
1480018100000000000     65648
1480018100000000000     1113037
1480018100000000000     19001
1480018110000000000     1113044
1480018110000000000     432433
1480018110000000000     2063570
1480018110000000000     827355
1480018110000000000     1009691
1480018110000000000     46671
1480018110000000000     3185160
1480018110000000000     47502
1480018110000000000     1764500
1480018110000000000     113732
1480018110000000000     65650
1480018120000000000     113734
1480018120000000000     65651
1480018120000000000     827357
1480018120000000000     2063641
1480018120000000000     1764513
1480018120000000000     46676
1480018120000000000     47505
1480018120000000000     3185293
1480018130000000000     47509
1480018130000000000     3185425
1480018130000000000     65652
1480018130000000000     432436
1480018130000000000     1113055
1480018130000000000     113736
1480018130000000000     340723
1480018130000000000     1009702
1480018130000000000     827358
1480018130000000000     136809
1480018130000000000     1764526
1480018130000000000     2063715
1480018140000000000     3185558
1480018140000000000     2063779
1480018140000000000     19011
1480018140000000000     47512
1480018140000000000     1113056
1480018140000000000     432437
1480018140000000000     136824
1480018140000000000     34732
1480018140000000000     432579
1480018140000000000     1009711
1480018150000000000     113742
1480018150000000000     1009712
1480018150000000000     1764548
1480018150000000000     65655
1480018150000000000     1113057
1480018150000000000     46710
1480018150000000000     827363
1480018150000000000     432440
1480018160000000000     827372
1480018160000000000     1009721
1480018160000000000     432584
1480018160000000000     3185825
1480018160000000000     47516
1480018160000000000     113745
1480018160000000000     19016
1480018160000000000     2063926
1480018160000000000     46710
1480018160000000000     34737
1480018160000000000     340728
1480018160000000000     136855
1480018160000000000     1113061
1480018170000000000     136869
1480018170000000000     113748
1480018170000000000     47518
1480018170000000000     432586
1480018170000000000     3185949
1480018170000000000     340729
1480018170000000000     2063992
1480018170000000000     1764586
1480018170000000000     19018
1480018170000000000     34739
1480018170000000000     65658
1480018170000000000     46710
1480018180000000000     65659
1480018180000000000     46715
1480018180000000000     1009731
1480018180000000000     432444
1480018180000000000     827385
1480018180000000000     113750
1480018180000000000     1113066
1480018180000000000     136887
1480018180000000000     34741
1480018180000000000     432591
1480018180000000000     19021
1480018180000000000     1764598
1480018190000000000     19024
1480018190000000000     1764607
1480018190000000000     2064133
1480018190000000000     3186227
1480018190000000000     46749
1480018190000000000     65660
1480018190000000000     47525
1480018190000000000     34745
1480018190000000000     113752
1480018190000000000     432595

> SELECT NON_NEGATIVE_DERIVATIVE(NumRecv, 10s) FROM metricname WHERE "a" = 'b' AND time > now() - 2m
name: metricname
time                    non_negative_derivative
----                    -----------------------
1480018100000000000     384933
1480018110000000000     680613
1480018140000000000     3.138049e+06
1480018160000000000     713630

> SELECT DERIVATIVE(NumRecv, 10s) FROM metricname WHERE "a" = 'b' AND time > now() - 2m
name: metricname
time                    derivative
----                    ----------
1480018100000000000     -298947
1480018110000000000     21
1480018120000000000     299141
1480018130000000000     -1.949905e+06
1480018140000000000     999320
1480018150000000000     651492
1480018160000000000     -754827
1480018170000000000     -990703
1480018180000000000     46641
1480018190000000000     -18910
1480018200000000000     -12002
```

I did the original testing on the 1.1 GA release, but upgraded our influx02 host and reproduced on 1.2.0~n201611240800

**Comment by daviesalex** [comment]
TL;DR this is operating as designed, but sucks. Will convert this to a feature request for an absolute value function!
